chicken-uninstall.1
.\" dummy line
.TH CHICKEN-UNINSTALL 1 "13 Aug 2008"

.SH NAME

chicken-uninstall \- uninstall extension library

.SH SYNOPSIS

chicken-uninstall [OPTION | PATTERN] ...

.SH DESCRIPTION

.I chicken\-uninstall
removes one or more already installed extension libraries for the CHICKEN
Scheme system.
.B PATTERN
may be a regular expression naming multiple extensions or just
an extension name.

.SH OPTIONS

Enter
.B chicken-uninstall\ \-help
for a list of supported command-line options.

.SH ENVIRONMENT\ VARIABLES

.TP
.B CHICKEN_PREFIX
The installation prefix where CHICKEN Scheme and its support files and
libraries are located. Defaults to the installation time prefix given
when configuring the system.

.TP
.B CHICKEN_INSTALL_PREFIX
An alternative installation prefix that will be prepended to extension
installation paths if specified. 

.TP
.B CHICKEN_REPOSITORY
The path where extension libraries are installed. Defaults to the package-library
path selected during configuration (usually
.B $prefix/lib/chicken/<binary\-version>
)


.SH DOCUMENTATION

More information can be found in the
.I CHICKEN\ User's\ Manual

.SH BUGS
Submit bug reports by e-mail to
.I chicken-janitors@nongnu.org
, preferrably using the
.B chicken\-bug
tool.

.SH AUTHORS
The CHICKEN Team

.SH SEE ALSO
.BR chicken-install(1)
.BR chicken-status(1)
.BR chicken(1)
.BR chicken-bug(1)


NEWS
4.5.2

- added `boot-chicken' makefile target to simplify bootstrapping
  the system from sources and documented this in the README
  (suggested by Jim Ursetto)
- CHICKEN can now be built on haiku (contributed by Chris Roberts)
- On Solaris, the system can be compiled with the SunPro C compiler
  (thanks to Semih Cemiloglu)
- removed the `-disable-warnings' compiler option and `disable-warnings'
  declaration specifier
- `fx/' and `fxmod' generate now faster code in safe mode
- cleaned up manual pages
- slightly optimized variable-value and procedure-access
- in the compiler `-debug-level 2' implies `-scrutinize'
- internal compiler-transformation for `for-each' and `map'
  apply now with any expression as the procedure argument
- the compiler warns about non-intrinsic exported toplevel variables
  which are declared to be safe
- `csc' didn't handle the `-verbose' option (thanks to Mario Goulart)
- the `,d' command in `csi' now detects circular lists (thanks to
  Christian Kellermann)
- strings passed to C runtime functions and which are converted to
  0-terminated C strings are now checked for not containing embedded
  `\0' characters (thanks to Peter Bex)
- errors in user-defined record printers are caught and shown in the 
  output instead of throwing an error to avoid endless recursion when
  an error message is printed
- a feature identifier named `chicken-MAJOR.MINOR' is now defined 
  to simplify conditionalization of code on the CHICKEN version
- `getter-with-setter' copies the lambda-information (if available) from
  the getter to the generated procedure
- `time' uses a somewhat more concise output format and outputs timing
  information to stderr instead of stdout
- added a new chapter on cross-development to the manual
- added the `safe-globals' declaration specifier
- split up manual chapter `Modules and macros' into two chapters
  (named `Modules' and `Macros', respectively - suggested by
  Mario Goulart)
- the last 5 non-precompiled regular expressions are now internally 
  cached in compiled form, which speeds up repeated matching of
  the same uncompiled regular expression dramatically
- added the new procedure `yes-or-no?' to the `utils' library unit
- added a `bench' makefile target that runs some non-trivial
  benchmark programs

4.5.1

- added `install-target' and `install-dev' makefile target for
  installing only target system libraries in cross-mode and
  development files (static libraries and headers)
- added `[-]no-procedure-checks-for-toplevel-bindings' compiler
  option and declaration
- usage of unimported syntax in modules gives more usable
  error messages; in particular, used but unimported FFI forms are 
  now detected
- invalid syntactic forms (mostly `()') encountered by the compiler
  or interpreter show the contextual form containing the expression,
  or, if indicated by the context warns about missing imports
- simplified manual pages of all core tools - they now refer to
  the output shown by invoking `<tool> -help'
- added new option `-feature FEATURE' to `chicken-install' tool
  to pass feature-identifiers to invocations of `csc'
- removed deprecated `-host-extension' option from `chicken-install'
- `chicken-status' in a system built for cross-compilation now 
  lists extensions installed in the target prefix, unless the
  new `-host' option is given on the command line
- `chicken-uninstall' in a system built for cross-compilation now 
  removes extensions installed in the target prefix, unless the
  new `-host' option is given on the command line
- added missing entry for `finite?' to the `chicken' module
  exports
- added new procedure `port-closed?' to the `library' unit
  (contributed by Peter Bex)
- added new procedure `symbol-append' to the library unit
- the compiler-option `-optimize-level 0' is equivalent to
  `-no-compiler-syntax -no-usual-integrations`
- internal rewritings of `map' and `for-each' ensure correct
  evaluation order of the arguments and does a better job
  at detecting non-list arguments (thanks to Jim Ursetto)
- `void' now takes arbitrary arguments and ignores them
- deprecated `noop' (from the `data-structures' unit) which
  is now replaced by `void'
- the `time' macro now performs a major garbage collection
  before executing the contained expressions and writes
  the timings in a more compact format to the port given
  by `(current-error-port)' instead of the standard output
  port
- definitions of the form `(define define ...)' and
  `(define-syntax define-syntax ...)' now trigger an error,
  as required by R5RS (thanks to Jeronimo Pellegrini and Alex
  Shinn)
- deprecated `random-seed' from the `extras' unit, since it
  is identical to `randomize'
- added new procedure `create-temporary-directory' to the
  `files' unit
- deprecated the optional path separator argument to
  `make-pathname'
- slightly improved the performance of keyword argument
  processing
- the second argument to `find-files' from the `posix' unit
  is now optional and defaults to `(constantly #t)'
- removed the deprecated `canonical-path' and `current-environment'
  procedures from the `posix' unit
- warnings that mostly refer to programming style issues are
  now coined `notes' and are only shown in the interpreter
  or when debug-mode is enabled or when scrutiny is enabled
  when compiling

4.5.0

- internal fixes of handling of alternative installation-prefix
  in setup-api
- certain compiler-warnings that are in really just notes
  and don't indicate a possible error (like reimport of
  identifiers) are only shown with -S or in verbose mode
- fixed handling of VARDIR in `chicken-install' (thanks to
  Davide Puricelli)
- `chicken-install -test' doesn't runs tests for dependencies
- when a non-else clause follows an else-clause in `cond', 
  `case' or `select' a warning (note) is shown in verbose
  mode
- removed the deprecated `define-extension' and
  `define-compiled-syntax'
- `chicken-uninstall' now always asks before removing 

chicken-uninstall.scm
;;;; chicken-uninstall.scm
;
; Copyright (c) 2008-2010, The Chicken Team
; All rights reserved.
;
; Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
; conditions are met:
;
;   Redistributions of source code must retain the above copyright notice, this list of conditions and the following
;     disclaimer. 
;   Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
;     disclaimer in the documentation and/or other materials provided with the distribution. 
;   Neither the name of the author nor the names of its contributors may be used to endorse or promote
;     products derived from this software without specific prior written permission. 
;
; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
; AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
; CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
; SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
; THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
; OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
; POSSIBILITY OF SUCH DAMAGE.


(require-library
 setup-api
 srfi-1 posix data-structures utils ports regex srfi-13 files)


(module main ()
  
  (import scheme chicken foreign)
  (import setup-api)
  (import srfi-1 posix data-structures utils ports regex srfi-13 files)

  (define-foreign-variable C_TARGET_LIB_HOME c-string)
  (define-foreign-variable C_BINARY_VERSION int)

  (define *host-extensions* #f)

  (define (repo-path)
    (if (and (feature? #:cross-chicken) (not *host-extensions*))
	(make-pathname C_TARGET_LIB_HOME (sprintf "chicken/~a" C_BINARY_VERSION))
	(repository-path)))

  (define *force* #f)

  (define (gather-eggs patterns)
    (let ((eggs (map pathname-file 
		     (glob (make-pathname (repo-path) "*" "setup-info")))))
      (delete-duplicates
       (concatenate (map (cut grep <> eggs) patterns))
       string=?)))

  (define (quit code)
    (print "aborted.")
    (exit code))

  (define (ask eggs)
    (handle-exceptions ex
	(if (eq? ex 'aborted)
	    (quit 1) 
	    (signal ex))
      (yes-or-no? 
       (string-concatenate
	(append
	 '("About to delete the following extensions:\n\n")
	 (map (cut string-append "  " <> "\n") eggs)
	 '("\nDo you want to proceed?")))
       default: "no"
       abort: abort-setup)))

  (define (uninstall pats)
    (let ((eggs (gather-eggs pats)))
      (cond ((null? eggs)
	     (print "nothing to remove.") )
	    ((or *force* (ask eggs))
	     (for-each
	      (lambda (e)
		(print "removing " e)
		(remove-extension e) )
	      eggs)))))

  (define (usage code)
    (print #<<EOF
usage: chicken-uninstall [OPTION | PATTERN] ...

  -h   -help                    show this message and exit
  -v   -version                 show version and exit
       -force                   don't ask, delete whatever matches
  -s   -sudo                    use sudo(1) for deleting files
       -host                    when cross-compiling, uninstall host extensions
EOF
);|
    (exit code))

  (define *short-options* '(#\h #\s))

  (define (main args)
    (let loop ((args args) (pats '()))
      (if (null? args)
	  (uninstall (if (null? pats) (usage 1) (reverse pats)))
	  (let ((arg (car args)))
	    (cond ((or (string=? arg "-help") 
		       (string=? arg "-h")
		       (string=? arg "--help"))
		   (usage 0))
		  ((or (string=? arg "-v") (string=? arg "-version"))
		   (print (chicken-version))
		   (exit 0))
		  ((string=? arg "-host")
		   (set! *host-extensions* #t)
		   (loop (cdr args) pats))
		  ((string=? arg "-force")
		   (set! *force* #t)
		   (loop (cdr args) pats))
		  ((or (string=? arg "-s") (string=? arg "-sudo"))
		   (sudo-install #t)
		   (loop (cdr args) pats))
		  ((and (positive? (string-length arg))
			(char=? #\- (string-ref arg 0)))
		   (if (> (string-length arg) 2)
		       (let ((sos (string->list (substring arg 1))))
			 (if (every (cut memq <> *short-options*) sos)
			     (loop (append (map (cut string #\- <>) sos) (cdr args)) pats)
			     (usage 1)))
		       (usage 1)))
		  (else (loop (cdr args) (cons (glob->regexp arg) pats))))))))

  (main (command-line-arguments))
  
 )
